example/bloc_library/lib/blocs/todos/todos_bloc.dart

```diff
@@ -0,0 +1,103 @@
+// Copyright 2018 The Flutter Architecture Sample Authors. All rights reserved.
+// Use of this source code is governed by the MIT license that can be found
+// in the LICENSE file.
+
+import 'package:bloc/bloc.dart';
+import 'package:meta/meta.dart';
+import 'package:bloc_library/blocs/todos/todos.dart';
+import 'package:bloc_library/models/models.dart';
+import 'package:todos_repository_flutter/todos_repository_flutter.dart';
+
+class TodosBloc extends Bloc<TodosEvent, TodosState> {
+  final TodosRepositoryFlutter todosRepository;
+
+  TodosBloc({@required this.todosRepository});
+
+  @override
+  TodosState get initialState => TodosLoading();
+
+  @override
+  Stream<TodosState> mapEventToState(
+    TodosState currentState,
+    TodosEvent event,
+  ) async* {
+    if (event is LoadTodos) {
+      yield* _mapLoadTodosToState();
+    } else if (event is AddTodo) {
+      yield* _mapAddTodoToState(currentState, event);
+    } else if (event is UpdateTodo) {
+      yield* _mapUpdateTodoToState(currentState, event);
+    } else if (event is DeleteTodo) {
+      yield* _mapDeleteTodoToState(currentState, event);
+    } else if (event is ToggleAll) {
+      yield* _mapToggleAllToState(currentState);
+    } else if (event is ClearCompleted) {
+      yield* _mapClearCompletedToState(currentState);
+    }
+  }
+
+  Stream<TodosState> _mapLoadTodosToState() async* {
+    try {
+      final todos = await this.todosRepository.loadTodos();
+      yield TodosLoaded(
+        todos.map(Todo.fromEntity).toList(),
+      );
+    } catch (_) {
+      yield TodosNotLoaded();
+    }
+  }
+
+  Stream<TodosState> _mapAddTodoToState(
+    TodosState currentState,
+    AddTodo event,
+  ) async* {
+    if (currentState is TodosLoaded) {
+      yield TodosLoaded(
+        List.from(currentState.todos)..add(event.todo),
+      );
+    }
+  }
+
+  Stream<TodosState> _mapUpdateTodoToState(
+    TodosState currentState,
+    UpdateTodo event,
+  ) async* {
+    if (currentState is TodosLoaded) {
+      yield TodosLoaded(
+        currentState.todos.map((todo) {
+          return todo.id == event.updatedTodo.id ? event.updatedTodo : todo;
+        }).toList(),
+      );
+    }
+  }
+
+  Stream<TodosState> _mapDeleteTodoToState(
```

Comment:
One error with this app: Please try mutating the todos then doing a Hot Restart. You will see the same todos that were initially loaded from the Mock API are returned once again.


In all of these operations that mutate the list of todos (add, remove, edit, clear completed, etc), you'll need to make a call out to the repository to save the changes.

Then, when the app is restarted, the repository should be up to date and return the correct current list of todos!

This is the one thing I haven't figured out how to integration test yet. Wonder if there's a newer method on the test driver for this!

Reply:
Oops! I’ll fix that asap 👍

Reply:
FYI: Until such feature is available on flutter_driver, it is possible to auto-restart an app via adb and re-run an integration test (or run another integration test) without reinstalling the app. Try the following for android:
```
#!/usr/bin/env bash
# Originally written by Maurice McCabe <mmcc007@gmail.com>, but placed in the public domain.

# Run Flutter integration tests on android device (or emulator)

show_help() {
    printf "\n\nusage: $0 package test

Utility for running integration tests for pre-installed flutter app on android device.
(app must be built in debug mode with 'enableFlutterDriverExtension()')

where:
    package
        name of package to run, eg, com.example.flutterapp
    test
        path of test to run, eg, test_driver/main_test.dart
"
    exit 1
}

run_test() {
    # note: assumes debug apk installed on device
    package_name=$1
    test_path=$2

    echo Starting Flutter app $package_name in debug mode...

    # stop app on device
    # (if running locally or started incorrectly by CI/CD)
    adb shell am force-stop $package_name

    # clear log (to avoid picking up any earlier observatory announcements on local re-runs)
    # (could comment out if running in CI/CD)
    adb logcat -c

    # start app on device
    adb shell am start -a android.intent.action.RUN -f 0x20000000 --ez enable-background-compilation true --ez enable-dart-profiling true --ez enable-checked-mode true $package_name/$package_name.MainActivity

    # wait for observatory startup on device and get port number
    obs_str=`adb logcat -e "Observatory listening on" -m 1`
    obs_port=`echo $obs_str | grep -Eo '([0-9]+)/$'`
    obs_port=${obs_port%?}
    echo Observatory on port $obs_port

    # forward a local port to observatory port on device
    forwarded_port=`adb forward tcp:0 tcp:$obs_port`
#    forwarded_port=4723 # re-use appium server port for now
    adb forward tcp:$forwarded_port tcp:$obs_port
    echo Local port $forwarded_port forwarded to observatory port $obs_port

    # run test
    echo Running integration test $test_path on app $package_name ...
    flutter packages get # required when running in CI/CD
    export VM_SERVICE_URL=http://127.0.0.1:$forwarded_port
    dart $test_path
#    flutter driver --use-existing-app=http://127.0.0.1:$forwarded_port --no-keep-app-running lib/main.dart

}

# if no arguments passed
if [[ -z $1  || -z $2 ]]; then show_help; fi

run_test $1 $2
```
Should be possible to do something similar for iOS, just haven't got it working yet...

Reply:
I posted an issue about flutter driver regarding this:
https://github.com/flutter/flutter/issues/28662